<a href="https://colab.research.google.com/github/anidhyabhatnagar/sttp1/blob/scientific_computing/Solving_Linear_Equations.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Notebook Authored By: <b>Anidhya Bhatnagar</b>
### Email: anidhya@gmail.com

# Solving Linear Equations

*   SciPy is a Python library, the collective name of the scientific computing
environment for Python, and the umbrella organization for many of the core libraries for scientific computing with Python.

*   The library, scipy, is in fact rather a collection of libraries for high-level scientific computing, which are more or less independent of each other.

*   The SciPy library is built on top of NumPy, which provides the basic array data structures and fundamental operations on such arrays. 

*   The modules in SciPy provide domain-specific high-level computation methods, such as routines for linear algebra, optimization, interpolation, integration, and much more.

*   The SciPy module `scipy` should be considered a collection of modules that are selectively imported when required. 

*   You can use the `scipy.linalg` module, for solving
linear systems of equations, and the `scipy.optimize` module, for solving nonlinear equations.

Importing libraries.

In [2]:
import numpy as np
from scipy import linalg as la

## Rank
*   Computing the *rank* of the matrix A that defines a linear equation system is a useful method that can tell us whether the matrix is singular or not and therefore whether there exists a solution or not.

*   When A has full rank, the solution is guaranteed to exist. 

*   However, it may or may not be possible to accurately compute the solution. 

## Condition Number

*   The *condition number* of the matrix, cond(A), gives a measure of how well or poorly conditioned a linear equation system is. 

*   If the conditioning number is close to 1, if the system is said to be well conditioned (a condition number 1 is ideal), and if the condition number is large, the system is said to be ill-conditioned. 

*   The solution to an equation system that is ill-conditioned can have large errors.

## Norm

*   The *norm* of a matrix is a measure of how large its elements are. 

*   It is a way of determining the “size” of a matrix that is not necessarily related to how many rows or columns the matrix has.

*   Matrix Norm or the norm of a matrix is a real number which is a measure of the magnitude of the matrix.

Assume that we have a linear equation system on the form `Ax = b`, where `x` is the solution vector. Now consider a small variation of `b`, say `δb`, which gives a corresponding change in the solution, `δx`, given by `A(x+δx) = b+δb`. 

Because of linearity of the equation, we have `Aδx = δb`. 

An important question to consider now is: *how large is the relative change in `x` compared to the relative change in `b`?* 

Mathematically we can formulate this question in terms of the ratios of the norms of these vectors.

Let's see it in action.

In [3]:
A = np.array([[2, 3], [5, 4]])
b = np.array([4, 3])

You can get the rank of the matrix using `linalg.matrix_rank` method, condition number by using `linalg.cond` method and norm by using `linalg.norm` method.

In [4]:
np.linalg.matrix_rank(A)

2

In [5]:
np.linalg.cond(A)

7.582401374401516

In [6]:
np.linalg.norm(A)

7.3484692283495345

In [13]:
np.linalg.solve(A, b)

array([-1.,  2.])

# LU Factorization

A better and computationally efficient method to solve this problem is LU Factorisation or Decomposition. 

The LU factorization of a matrix A is such that `A = LU`, where L is lower triangular matrix and U is upper triangular matrix.

Given L and U the solution vector x can be efficiently constructed by first solving `Ly = b` with forward substitution and then solving `Ux = y` with backward substitution.

You can use `la.lu` method in `scipy`'s linear algebra module.

In [8]:
A = np.array([[2, 3], [5, 4]])
b = np.array([4, 3])
P, L, U = la.lu(A)

In [9]:
L 

array([[1. , 0. ],
       [0.4, 1. ]])

In [10]:
U

array([[5. , 4. ],
       [0. , 1.4]])

In [11]:
P.dot(L.dot(U))

array([[2., 3.],
       [5., 4.]])

In [12]:
la.solve(A, b)

array([-1.,  2.])



---



*The above methods are for Square Systems where `m = n`. However, the Rectangular Systems are where `m ≠ n`, and hence, they are either underdetermined or overdetermined.*

*Underdetermined systems have more variables than equations, so the solution cannot be fully determined. Therefore, for such a system, the solution must be given in terms of the remaining free variables. This makes it difficult to treat this type of problem numerically, but a symbolic approach can often be used instead.*

*The SciPy uses `sympy` library for solving such problems using symbolic approach.* 



---



# Optimization

*   In general, optimization is the process of finding and selecting the optimal element from a set of feasible candidates. 

*   In mathematical optimization, this problem is usually formulated as determining the extreme value of a function on a given domain. 

*   An extreme value, or an optimal value, can refer to either the minimum or maximum of the function, depending on the application and the specific problem.

*   Optimization is closely related to equation solving because at an optimal value of a function, its derivative, or gradient in the multivariate case, is zero.

*   SciPy has an optimization module `optimize` for solving non-linear optimization problems.

*   However, convex optimization library `cvxopt` can be used for solving linear optimization problems with linear constraints and it also provides solvers for quadratic optimization problems.





---



Lets take the following optimization problem:

> **Minimize the area of a cylinder with unit volume.**

> Here, suitable variables are the radius `r` and height `h` of the cylinder, and the objective function is `f([r, h]) = 2πr^2 + 2πrh`, subject to the equality constraint `g([r, h]) = πr^2h − 1 = 0`. 

This problem is a two-dimensional optimization problem with an equality constraint.



To solve this problem using SciPy’s numerical optimization functions, 

*   First define a Python function `f` that implements the objective function. 

*   To solve the optimization problem, we then pass this function to, for example, `optimize.brent`. 

*  Optionally we can use the `brack` keyword argument to specify a starting interval for the algorithm.

In [14]:
from scipy import optimize

In [17]:
def f(r):
  return 2 * np.pi * r**2 + 2 / r

In [18]:
r_min = optimize.brent(f, brack=(0.1, 4))
r_min

0.5419260772557135

In [19]:
f(r_min)

5.535810445932086

Instead of calling `optimize.brent` directly, you could use the generic interface for scalar minimization problems `optimize.minimize_scalar`. 

> *Note that to specify a starting interval in this case, you must use the* `bracket` *keyword argument*

In [21]:
optimize.minimize_scalar(f, bracket=(0.1, 4))

     fun: 5.535810445932086
    nfev: 19
     nit: 15
 success: True
       x: 0.5419260772557135

> Both the methods above give that the radius that minimizes the area of the cylinder is approximately `0.54` and a minimum area of approximately `5.54`.



---

